File: nodejs-typescript/template/src/sample/sample-document.service.ts
{{- if eq stack_profile "nosql-cache" }}
import { Injectable, NotFoundException } from '@nestjs/common';
import {
  SampleDocument,
  SampleDocumentRepository,
} from './sample-document.repository';
import { SampleCache } from '../cache/sample-cache.service';

@Injectable()
export class SampleDocumentService {
  constructor(
    private readonly repository: SampleDocumentRepository,
    private readonly cache: SampleCache,
  ) {}

  async getByName(name: string): Promise<SampleDocument> {
    const cached = await this.cache.get(name);
    if (cached) return JSON.parse(cached) as SampleDocument;

    const doc = await this.repository.findByName(name);
    if (!doc) throw new NotFoundException(`SampleDocument ${name} not found`);
    await this.cache.put(name, JSON.stringify(doc));
    return doc;
  }

  async create(input: { name: string; payload: string }): Promise<SampleDocument> {
    const created = await this.repository.insert(input);
    await this.cache.put(created.name, JSON.stringify(created));
    return created;
  }
}
{{- end }}

File: kotlin-microservice/template/src/main/kotlin/{{tld}}/{{author}}/{{app_name}}/service/SampleDocumentService.kt
{{- if eq stack_profile "nosql-cache" }}
package {{tld}}.{{author}}.{{app_name}}.service

import {{tld}}.{{author}}.{{app_name}}.cache.SampleCache
import {{tld}}.{{author}}.{{app_name}}.persistence.SampleDocument
import {{tld}}.{{author}}.{{app_name}}.persistence.SampleDocumentRepository
import org.springframework.stereotype.Service

@Service
class SampleDocumentService(
    private val repository: SampleDocumentRepository,
    private val cache: SampleCache,
) {
    fun insert(doc: SampleDocument) {
        repository.insert(doc)
    }

    fun getByName(name: String): SampleDocument? {
        // Cache stores only the id (StringRedisTemplate-backed). If SampleDocument grows
        // new fields, either add JSON serialization here or switch to RedisTemplate<String, SampleDocument>.
        val cacheKey = "by-name:$name"
        cache.get(cacheKey)?.let { cachedId ->
            return SampleDocument(id = cachedId, name = name)
        }
        val doc = repository.findByName(name) ?: return null
        cache.put(cacheKey, doc.id)
        return doc
    }

    fun evict(name: String) {
        cache.evict("by-name:$name")
    }
}
{{- end }}
